### ISSUE TITLE
[BUG] Deterministic variables when restarting from checkpoint

### ISSUE THREAD

**Issue by davidcerny** [revbayes/revbayes#339]
**Describe the bug**
I performed a partitioned, topologically constrained FBD analysis using the MPI version of RevBayes, running 10 chains in parallel and saving checkpoint files every 100 iterations. To get around issue #184, I then used the single-thread version of RevBayes to restart each of the 10 runs individually. The analysis made one attempt to draw an initial state and then threw the following error, originating from the `setValueFromString()` member function of class `DeterministicNode`: 

```
Error:	Cannot set a deterministic node from a string.
```

I tried to investigate which variable was responsible for this, since the overall model contained a number of deterministic nodes. However, when I recompiled the exact same version of RevBayes with `-debug true` and ran it within GDB, I was unable to reproduce the issue, as the analysis instead threw a different error:

```
Assertion failed: (isnan(sum) || (0 <= sum and sum <= 1.00000000001)), function computeInternalNodeLikelihood, file PhyloCTMCSiteHomogeneous.h, line 264.
```

**To Reproduce**
A zip file with the script, data files, and the checkpoint state file is available [here](https://github.com/revbayes/revbayes/files/11056197/detnode_restart_error_example.zip). One absolute path has to be changed in the script.

**Expected behavior**
The analysis should be successfully re-initialized. I would expect RevBayes to re-initialize deterministic nodes conditional on the current values of their parent nodes, instead of attempting to set their values from a string.

**Screenshots**
N/A

**Computer info**
I encountered this problem with both the stable (rapture-1925-g7b872a) and development (rapture-2131-g72b6d3) versions of RevBayes 1.2.1 on Red Hat Enterprise Linux 8.4 and macOS 12.6.3.

**Comment by jsigao** [comment]
Hi David, after glancing over the checkpoint file (`zhang_et_wang_19_rjMCMC_partially_constrained_run_1.state`), I think this may be related to a similar checkpointing issue that I encountered and tried to fix a short while ago. Specifically, the issue is caused by the mismatch between the ways that a stochastic-variable vector is written to and read from the checkpoint file. 

For example, in your analysis each element (e.g., `rel_branch_rates[1]`) of the `rel_branch_rates` vector is a stochastic variable and also a vector itself. Then [the current implementation](https://github.com/revbayes/revbayes/blob/8b35be91f664855621e5633d5433714a836af800/src/core/analysis/mcmc/Mcmc.cpp#L205) ''flattens'' that vector when writes to the checkpoint file so that `rel_branch_rates[1]` and `rel_branch_rates[2]` (say `rel_branch_rates[1]=[1.0,1.1]` and `rel_branch_rates[2]=[1.2,1.3]`) will be logged as:
```
rel_branch_rates[1][1]\trel_branch_rates[1][2]\trel_branch_rates[2][1]\trel_branch_rates[2][2]
1.0\t1.1\t1.2\t1.3
```

However, when reading in the checkpoint file, I believe [the current implementation](https://github.com/revbayes/revbayes/blob/8b35be91f664855621e5633d5433714a836af800/src/core/analysis/mcmc/Mcmc.cpp#L750) assumes that each element separated by the delimiter `\t` is an independent variable (i.e., it doesn't understand the relationship between the flattened elements `rel_branch_rates[1][1]` and `rel_branch_rates[1][2]`) so it cannot reassemble `rel_branch_rates[1]` from its flattened elements. 

Changing [this line](https://github.com/revbayes/revbayes/blob/8b35be91f664855621e5633d5433714a836af800/src/core/analysis/mcmc/Mcmc.cpp#L205) to `bool flatten = false;` likely will fix the problem (as what I did in [this commit](https://github.com/revbayes/revbayes/commit/ec20e7875f3f572c7ad55b0767b40c7492538b4c); I haven't been able to submit this fix as a PR to the development branch since I haven't found enough time to do some more thorough checks with it). After that fix `rel_branch_rates[1]` will be logged as a single vectorized element. That is, it will be logged as:
```
rel_branch_rates[1]\trel_branch_rates[2]
[1.0,1.1]\t[1.2,1.3]
```
which matches the way how a stochastic-variable vector is read in from a checkpoint file.

There could be a better fix to this particular issue or other issues with the checkpointing functions underlying your analysis; I'll leave that to other developers who are more familiar with the checkpointing module.

**Comment by davidcerny** [comment]
Thanks a lot, Jiansi! You are 100% right; recompiling RevBayes after changing that single line in `Mcmc.cpp` was enough to solve the issue.

> I haven't been able to submit this fix as a PR to the development branch since I haven't found enough time to do some more thorough checks with it

If there are any particular checks you have in mind that I could potentially help run, I'd be happy to do so. It would be great to get this fix into the next minor release.

**Comment by jsigao** [comment]
Hi David, glad to hear that it solves the issue! I had checked with my own analysis that after the fixes I made (including this one and a few others) the checkpointing function seemed to work properly (both with and without MPI); I guess it was just that I felt there were still something that I didn't fully understand about the checkpointing code and that checkpoint still doesn't work properly under MC^3, so I didn't feel quite ready for submitting the fixes as a PR. 

However, as you suggested, the current fix may already be able to solve some of the checkpointing issues, so I'll submit a PR this week with the fixes that I'm more confident about (including this one as now there have been at least two test points) to accelerate the process. I'll suggest you and some other developers who have worked on the checkpointing functions as reviewers for that PR.
